File: day_18_css_box_model/day_18_css_box_model.js
"use strict";

/*
 * CSS Box Model: practical JavaScript companion
 *
 * This file demonstrates:
 *   - CSS box-model calculations
 *   - content-box and border-box
 *   - padding, border, and margin
 *   - overflow detection
 *   - DOM measurement
 *   - getComputedStyle()
 *   - offsetWidth/clientWidth/scrollWidth
 *   - validation
 *   - responsive calculations
 *   - a small reusable BoxModelInspector
 *
 * Run in a browser by loading this file from an HTML page.
 * The computational examples also work in Node.js.
 */

// ============================================================================
// 1. BASIC BOX MODEL CALCULATIONS
// ============================================================================

function calculateContentBoxWidth(contentWidth) {
    return contentWidth;
}

function calculatePaddingBoxWidth(contentWidth, paddingLeft, paddingRight) {
    return contentWidth + paddingLeft + paddingRight;
}

function calculateBorderBoxWidth(
    contentWidth,
    paddingLeft,
    paddingRight,
    borderLeft,
    borderRight
) {
    return (
        contentWidth +
        paddingLeft +
        paddingRight +
        borderLeft +
        borderRight
    );
}

function calculateMarginBoxWidth(
    contentWidth,
    paddingLeft,
    paddingRight,
    borderLeft,
    borderRight,
    marginLeft,
    marginRight
) {
    return (
        calculateBorderBoxWidth(
            contentWidth,
            paddingLeft,
            paddingRight,
            borderLeft,
            borderRight
        ) +
        marginLeft +
        marginRight
    );
}

console.log("=== Basic box model calculation ===");

const contentWidth = 200;
const paddingLeft = 20;
const paddingRight = 20;
const borderLeft = 5;
const borderRight = 5;
const marginLeft = 30;
const marginRight = 30;

console.log("Content box:", calculateContentBoxWidth(contentWidth));
console.log(
    "Padding box:",
    calculatePaddingBoxWidth(contentWidth, paddingLeft, paddingRight)
);
console.log(
    "Border box:",
    calculateBorderBoxWidth(
        contentWidth,
        paddingLeft,
        paddingRight,
        borderLeft,
        borderRight
    )
);
console.log(
    "Margin box:",
    calculateMarginBoxWidth(
        contentWidth,
        paddingLeft,
        paddingRight,
        borderLeft,
        borderRight,
        marginLeft,
        marginRight
    )
);


// ============================================================================
// 2. content-box AND border-box
// ============================================================================

function calculateUsedBorderBoxWidth({
    declaredWidth,
    paddingLeft,
    paddingRight,
    borderLeft,
    borderRight,
    boxSizing
}) {
    if (declaredWidth < 0) {
        throw new RangeError("Width cannot be negative.");
    }

    if (boxSizing === "content-box") {
        return (
            declaredWidth +
            paddingLeft +
            paddingRight +
            borderLeft +
            borderRight
        );
    }

    if (boxSizing === "border-box") {
        return declaredWidth;
    }

    throw new Error(
        "boxSizing must be 'content-box' or 'border-box'."
    );
}

const sizingExample = {
    declaredWidth: 200,
    paddingLeft: 20,
    paddingRight: 20,
    borderLeft: 5,
    borderRight: 5
};

console.log("\n=== box-sizing ===");

console.log(
    "content-box:",
    calculateUsedBorderBoxWidth({
        ...sizingExample,
        boxSizing: "content-box"
    })
);

console.log(
    "border-box:",
    calculateUsedBorderBoxWidth({
        ...sizingExample,
        boxSizing: "border-box"
    })
);


// ============================================================================
// 3. VALIDATION
// ============================================================================

function validateBoxValues(values) {
    const errors = [];

    const nonNegativeProperties = [
        "width",
        "height",
        "padding",
        "border"
    ];

    for (const property of nonNegativeProperties) {
        if (values[property] < 0) {
            errors.push(`${property} cannot be negative.`);
        }
    }

    // Margin is deliberately excluded because negative margins are valid CSS.
    return errors;
}

console.log("\n=== Validation ===");

console.log(
    validateBoxValues({
        width: 300,
        height: 150,
        padding: 20,
        border: 1,
        margin: -10
    })
);

console.log(
    validateBoxValues({
        width: -1,
        height: 150,
        padding: -5,
        border: 1,
        margin: 10
    })
);


// ============================================================================
// 4. OVERFLOW CALCULATION
// ============================================================================

function calculateOverflow(contentWidth, contentHeight, boxWidth, boxHeight) {
    return {
        horizontal: Math.max(0, contentWidth - boxWidth),
        vertical: Math.max(0, contentHeight - boxHeight)
    };
}

console.log("\n=== Overflow ===");

console.log(
    calculateOverflow(
        600,
        500,
        400,
        300
    )
);


// ============================================================================
// 5. RESPONSIVE WIDTH
// ============================================================================

function clamp(value, minimum, maximum) {
    return Math.min(Math.max(value, minimum), maximum);
}

function responsiveComponentWidth(
    viewportWidth,
    percentage,
    minimum,
    maximum
) {
    return clamp(
        viewportWidth * percentage,
        minimum,
        maximum
    );
}

console.log("\n=== Responsive sizing ===");

for (const viewportWidth of [320, 480, 768, 1024, 1440]) {
    console.log(
        `${viewportWidth}px viewport -> ` +
        `${responsiveComponentWidth(viewportWidth, 0.8, 280, 900)}px`
    );
}


// ============================================================================
// 6. DOM DEMONSTRATION
// ============================================================================

function createBoxModelDemo() {
    if (typeof document === "undefined") {
        console.log(
            "\nDOM demonstration skipped because this runtime has no document."
        );
        return;
    }

    const container = document.createElement("section");
    container.id = "box-model-demo";

    const title = document.createElement("h2");
    title.textContent = "CSS Box Model Demo";

    const box = document.createElement("div");

    box.textContent =
        "This element contains content, padding, border, and margin.";

    Object.assign(box.style, {
        width: "240px",
        height: "120px",
        padding: "20px",
        border: "4px solid black",
        margin: "30px",
        boxSizing: "content-box",
        overflow: "auto",
        overflowWrap: "anywhere"
    });

    container.append(title, box);
    document.body.appendChild(container);

    return box;
}


// ============================================================================
// 7. DOM MEASUREMENT
// ============================================================================

function inspectElementBox(element) {
    if (!(element instanceof Element)) {
        throw new TypeError("Expected a DOM Element.");
    }

    const computed = getComputedStyle(element);

    const measurement = {
        width: computed.width,
        height: computed.height,

        paddingTop: computed.paddingTop,
        paddingRight: computed.paddingRight,
        paddingBottom: computed.paddingBottom,
        paddingLeft: computed.paddingLeft,

        borderTopWidth: computed.borderTopWidth,
        borderRightWidth: computed.borderRightWidth,
        borderBottomWidth: computed.borderBottomWidth,
        borderLeftWidth: computed.borderLeftWidth,

        marginTop: computed.marginTop,
        marginRight: computed.marginRight,
        marginBottom: computed.marginBottom,
        marginLeft: computed.marginLeft,

        boxSizing: computed.boxSizing,

        clientWidth: element.clientWidth,
        clientHeight: element.clientHeight,

        offsetWidth: element.offsetWidth,
        offsetHeight: element.offsetHeight,

        scrollWidth: element.scrollWidth,
        scrollHeight: element.scrollHeight
    };

    return measurement;
}


// ============================================================================
// 8. UNDERSTANDING DOM DIMENSION PROPERTIES
// ============================================================================

function explainDomDimensions() {
    console.log(`
DOM dimension concepts:

clientWidth/clientHeight:
    Content + padding. Borders and most scrollbar space are excluded.

offsetWidth/offsetHeight:
    Border-box dimensions, generally including scrollbar space.

scrollWidth/scrollHeight:
    Size required to contain the element's content, including content
    that is not currently visible because of overflow.

getComputedStyle(element):
    Provides computed CSS values, useful for inspecting padding, borders,
    margins, dimensions, overflow, and box-sizing.
`);
}


// ============================================================================
// 9. DETECTING OVERFLOW IN THE BROWSER
// ============================================================================

function hasHorizontalOverflow(element) {
    return element.scrollWidth > element.clientWidth;
}

function hasVerticalOverflow(element) {
    return element.scrollHeight > element.clientHeight;
}

function getOverflowState(element) {
    return {
        horizontal: hasHorizontalOverflow(element),
        vertical: hasVerticalOverflow(element)
    };
}


// ============================================================================
// 10. REUSABLE INSPECTOR CLASS
// ============================================================================

class BoxModelInspector {
    constructor(element) {
        if (!(element instanceof Element)) {
            throw new TypeError("BoxModelInspector requires a DOM Element.");
        }

        this.element = element;
    }

    computedStyle() {
        return getComputedStyle(this.element);
    }

    dimensions() {
        const style = this.computedStyle();

        return {
            declaredWidth: style.width,
            declaredHeight: style.height,
            boxSizing: style.boxSizing,

            clientWidth: this.element.clientWidth,
            clientHeight: this.element.clientHeight,

            offsetWidth: this.element.offsetWidth,
            offsetHeight: this.element.offsetHeight,

            scrollWidth: this.element.scrollWidth,
            scrollHeight: this.element.scrollHeight
        };
    }

    spacing() {
        const style = this.computedStyle();

        return {
            padding: {
                top: style.paddingTop,
                right: style.paddingRight,
                bottom: style.paddingBottom,
                left: style.paddingLeft
            },
            border: {
                top: style.borderTopWidth,
                right: style.borderRightWidth,
                bottom: style.borderBottomWidth,
                left: style.borderLeftWidth
            },
            margin: {
                top: style.marginTop,
                right: style.marginRight,
                bottom: style.marginBottom,
                left: style.marginLeft
            }
        };
    }

    overflow() {
        return {
            horizontal: hasHorizontalOverflow(this.element),
            vertical: hasVerticalOverflow(this.element)
        };
    }

    report() {
        return {
            dimensions: this.dimensions(),
            spacing: this.spacing(),
            overflow: this.overflow()
        };
    }
}


// ============================================================================
// 11. RESPONSIVE BOX USING CSS CUSTOM PROPERTIES
// ============================================================================

function createResponsiveCard() {
    if (typeof document === "undefined") {
        return null;
    }

    const card = document.createElement("article");

    card.textContent =
        "Responsive card with predictable box sizing and overflow handling.";

    card.style.setProperty("box-sizing", "border-box");
    card.style.setProperty("width", "min(100%, 360px)");
    card.style.setProperty("padding", "24px");
    card.style.setProperty("border", "1px solid currentColor");
    card.style.setProperty("margin", "16px");
    card.style.setProperty("overflow", "auto");
    card.style.setProperty("overflow-wrap", "anywhere");

    document.body.appendChild(card);

    return card;
}


// ============================================================================
// 12. EDGE CASE: LONG UNBREAKABLE CONTENT
// ============================================================================

function demonstrateLongContent() {
    if (typeof document === "undefined") {
        return;
    }

    const container = document.createElement("div");
    container.style.width = "300px";
    container.style.boxSizing = "border-box";
    container.style.border = "1px solid black";
    container.style.padding = "12px";

    const content = document.createElement("div");
    content.textContent =
        "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa";

    // Without an appropriate wrapping rule, long tokens can produce
    // unexpected horizontal overflow.
    content.style.overflowWrap = "anywhere";

    container.appendChild(content);
    document.body.appendChild(container);

    return container;
}


// ============================================================================
// 13. MARGIN AND PADDING COMPARISON
// ============================================================================

function compareSpacing() {
    return {
        padding: {
            insideBorder: true,
            partOfBorderBox: true,
            negativeAllowed: false
        },
        margin: {
            outsideBorder: true,
            partOfBorderBox: false,
            negativeAllowed: true
        }
    };
}

console.log("\n=== Padding versus margin ===");
console.log(compareSpacing());


// ============================================================================
// 14. SIMPLE ASSERTIONS
// ============================================================================

function assertEqual(actual, expected, description) {
    if (actual !== expected) {
        throw new Error(
            `${description}: expected ${expected}, received ${actual}`
        );
    }
}

function runTests() {
    assertEqual(
        calculateBorderBoxWidth(200, 20, 20, 5, 5),
        250,
        "content-box width calculation"
    );

    assertEqual(
        calculateUsedBorderBoxWidth({
            declaredWidth: 200,
            paddingLeft: 20,
            paddingRight: 20,
            borderLeft: 5,
            borderRight: 5,
            boxSizing: "border-box"
        }),
        200,
        "border-box width calculation"
    );

    const overflow = calculateOverflow(600, 500, 400, 300);

    assertEqual(
        overflow.horizontal,
        200,
        "horizontal overflow calculation"
    );

    assertEqual(
        overflow.vertical,
        200,
        "vertical overflow calculation"
    );

    console.log("\nAll JavaScript tests passed.");
}


// ============================================================================
// 15. BROWSER EXECUTION
// ============================================================================

function runBrowserDemonstration() {
    if (typeof document === "undefined") {
        return;
    }

    const box = createBoxModelDemo();
    createResponsiveCard();
    demonstrateLongContent();

    if (box) {
        const inspector = new BoxModelInspector(box);

        console.log("\n=== Browser box-model inspection ===");
        console.table(inspector.report());
        console.log("Overflow:", inspector.overflow());
    }

    explainDomDimensions();
}


// ============================================================================
// 16. START
// ============================================================================

runTests();
runBrowserDemonstration();

File: day_18_css_box_model/day_18_css_box_model.cpp
#include <algorithm>
#include <cmath>
#include <iomanip>
#include <iostream>
#include <limits>
#include <map>
#include <optional>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

/*
 * CSS Box Model Case Study
 *
 * Scenario:
 *     A web application needs a predictable component sizing engine for
 *     cards, panels, and responsive containers. The C++ program models
 *     important CSS box-model calculations so the layout rules can be
 *     tested independently of a browser.
 *
 * The program demonstrates:
 *     content
 *     padding
 *     border
 *     margin
 *     content-box
 *     border-box
 *     dimensions
 *     min/max constraints
 *     overflow
 *     validation
 *     reusable classes
 *     test-like verification
 *
 * Compile:
 *     C++17 or later
 */

namespace cssbox {

// ============================================================================
// Basic types
// ============================================================================

struct Edges {
    double top{0};
    double right{0};
    double bottom{0};
    double left{0};

    double horizontal() const {
        return left + right;
    }

    double vertical() const {
        return top + bottom;
    }
};

enum class BoxSizing {
    ContentBox,
    BorderBox
};

enum class OverflowMode {
    Visible,
    Hidden,
    Clip,
    Scroll,
    Auto
};

struct Size {
    double width{0};
    double height{0};
};

struct OverflowResult {
    double horizontal{0};
    double vertical{0};

    bool hasHorizontalOverflow() const {
        return horizontal > 0;
    }

    bool hasVerticalOverflow() const {
        return vertical > 0;
    }
};

// ============================================================================
// Utility validation
// ============================================================================

void requireNonNegative(double value, const std::string& field) {
    if (!std::isfinite(value)) {
        throw std::invalid_argument(field + " must be finite.");
    }

    if (value < 0) {
        throw std::invalid_argument(field + " cannot be negative.");
    }
}

void validateEdges(const Edges& edges, const std::string& field) {
    requireNonNegative(edges.top, field + ".top");
    requireNonNegative(edges.right, field + ".right");
    requireNonNegative(edges.bottom, field + ".bottom");
    requireNonNegative(edges.left, field + ".left");
}

// ============================================================================
// Box model
// ============================================================================

class BoxModel {
private:
    Size declaredSize_;
    Edges padding_;
    Edges border_;
    Edges margin_;
    BoxSizing boxSizing_;

    std::optional<double> minWidth_;
    std::optional<double> maxWidth_;
    std::optional<double> minHeight_;
    std::optional<double> maxHeight_;

public:
    BoxModel(
        Size declaredSize,
        Edges padding,
        Edges border,
        Edges margin,
        BoxSizing boxSizing = BoxSizing::ContentBox
    )
        : declaredSize_(declaredSize),
          padding_(padding),
          border_(border),
          margin_(margin),
          boxSizing_(boxSizing) {

        requireNonNegative(declaredSize_.width, "declared width");
        requireNonNegative(declaredSize_.height, "declared height");
        validateEdges(padding_, "padding");
        validateEdges(border_, "border");

        // Margins are intentionally not validated as non-negative because
        // CSS permits negative margins.
        if (!std::isfinite(margin_.top) ||
            !std::isfinite(margin_.right) ||
            !std::isfinite(margin_.bottom) ||
            !std::isfinite(margin_.left)) {
            throw std::invalid_argument("Margins must be finite.");
        }
    }

    void setWidthConstraints(
        std::optional<double> minimum,
        std::optional<double> maximum
    ) {
        if (minimum && *minimum < 0) {
            throw std::invalid_argument("Minimum width cannot be negative.");
        }

        if (maximum && *maximum < 0) {
            throw std::invalid_argument("Maximum width cannot be negative.");
        }

        if (minimum && maximum && *minimum > *maximum) {
            throw std::invalid_argument(
                "Minimum width cannot exceed maximum width."
            );
        }

        minWidth_ = minimum;
        maxWidth_ = maximum;
    }

    void setHeightConstraints(
        std::optional<double> minimum,
        std::optional<double> maximum
    ) {
        if (minimum && *minimum < 0) {
            throw std::invalid_argument("Minimum height cannot be negative.");
        }

        if (maximum && *maximum < 0) {
            throw std::invalid_argument("Maximum height cannot be negative.");
        }

        if (minimum && maximum && *minimum > *maximum) {
            throw std::invalid_argument(
                "Minimum height cannot exceed maximum height."
            );
        }

        minHeight_ = minimum;
        maxHeight_ = maximum;
    }

    Size contentSize() const {
        if (boxSizing_ == BoxSizing::ContentBox) {
            return declaredSize_;
        }

        // border-box means declared dimensions include padding and border.
        const double contentWidth =
            declaredSize_.width -
            padding_.horizontal() -
            border_.horizontal();

        const double contentHeight =
            declaredSize_.height -
            padding_.vertical() -
            border_.vertical();

        if (contentWidth < 0 || contentHeight < 0) {
            throw std::invalid_argument(
                "Border-box dimensions are too small for "
                "the specified padding and border."
            );
        }

        return {contentWidth, contentHeight};
    }

    Size paddingBoxSize() const {
        const Size content = contentSize();

        return {
            content.width + padding_.horizontal(),
            content.height + padding_.vertical()
        };
    }

    Size borderBoxSize() const {
        const Size paddingBox = paddingBoxSize();

        return {
            paddingBox.width + border_.horizontal(),
            paddingBox.height + border_.vertical()
        };
    }

    Size constrainedBorderBoxSize() const {
        Size result = borderBoxSize();

        if (minWidth_) {
            result.width = std::max(result.width, *minWidth_);
        }

        if (maxWidth_) {
            result.width = std::min(result.width, *maxWidth_);
        }

        if (minHeight_) {
            result.height = std::max(result.height, *minHeight_);
        }

        if (maxHeight_) {
            result.height = std::min(result.height, *maxHeight_);
        }

        return result;
    }

    Size marginBoxSize() const {
        const Size borderBox = constrainedBorderBoxSize();

        return {
            borderBox.width + margin_.horizontal(),
            borderBox.height + margin_.vertical()
        };
    }

    const Edges& padding() const {
        return padding_;
    }

    const Edges& border() const {
        return border_;
    }

    const Edges& margin() const {
        return margin_;
    }

    BoxSizing boxSizing() const {
        return boxSizing_;
    }
};

// ============================================================================
// Overflow
// ============================================================================

OverflowResult calculateOverflow(
    Size contentSize,
    Size availableSize
) {
    if (contentSize.width < 0 ||
        contentSize.height < 0 ||
        availableSize.width < 0 ||
        availableSize.height < 0) {
        throw std::invalid_argument(
            "Content and available dimensions cannot be negative."
        );
    }

    return {
        std::max(0.0, contentSize.width - availableSize.width),
        std::max(0.0, contentSize.height - availableSize.height)
    };
}

std::string overflowDescription(
    OverflowMode mode
) {
    switch (mode) {
        case OverflowMode::Visible:
            return "visible: content may extend outside the box";

        case OverflowMode::Hidden:
            return "hidden: overflowing content is clipped";

        case OverflowMode::Clip:
            return "clip: content is clipped without creating scrolling";

        case OverflowMode::Scroll:
            return "scroll: scrolling mechanism is requested";

        case OverflowMode::Auto:
            return "auto: scrolling is provided when needed";
    }

    return "unknown";
}

// ============================================================================
// Responsive width model
// ============================================================================

double responsiveWidth(
    double containerWidth,
    double percentage,
    double minimum,
    double maximum
) {
    requireNonNegative(containerWidth, "container width");
    requireNonNegative(percentage, "percentage");
    requireNonNegative(minimum, "minimum");
    requireNonNegative(maximum, "maximum");

    if (minimum > maximum) {
        throw std::invalid_argument(
            "Minimum responsive width cannot exceed maximum width."
        );
    }

    const double preferred = containerWidth * percentage;

    return std::clamp(preferred, minimum, maximum);
}

// ============================================================================
// Margin collapse demonstration
// ============================================================================

double collapsePositiveVerticalMargins(
    double first,
    double second
) {
    requireNonNegative(first, "first margin");
    requireNonNegative(second, "second margin");

    // This is a simplified representation of two adjoining positive
    // vertical margins in normal block flow.
    return std::max(first, second);
}

// ============================================================================
// Application-level component
// ============================================================================

class ProductCard {
private:
    std::string productName_;
    double viewportWidth_;
    BoxModel box_;

public:
    ProductCard(
        std::string productName,
        double viewportWidth
    )
        : productName_(std::move(productName)),
          viewportWidth_(viewportWidth),
          box_(
              {360, 220},
              {24, 24, 24, 24},
              {1, 1, 1, 1},
              {16, 16, 16, 16},
              BoxSizing::BorderBox
          ) {

        requireNonNegative(viewportWidth_, "viewport width");

        // The card cannot exceed its available viewport width.
        box_.setWidthConstraints(
            280.0,
            std::min(360.0, viewportWidth_)
        );
    }

    void printReport() const {
        const Size content = box_.contentSize();
        const Size padding = box_.paddingBoxSize();
        const Size border = box_.constrainedBorderBoxSize();
        const Size margin = box_.marginBoxSize();

        std::cout << "\nProduct card: " << productName_ << "\n";
        std::cout << "----------------------------------------\n";

        std::cout << "Content box:  "
                  << content.width << " x "
                  << content.height << " px\n";

        std::cout << "Padding box:  "
                  << padding.width << " x "
                  << padding.height << " px\n";

        std::cout << "Border box:   "
                  << border.width << " x "
                  << border.height << " px\n";

        std::cout << "Margin box:   "
                  << margin.width << " x "
                  << margin.height << " px\n";
    }
};

// ============================================================================
// Demonstrations
// ============================================================================

void demonstrateBasicBoxModel() {
    std::cout << "\n=== BASIC BOX MODEL ===\n";

    BoxModel box(
        {200, 100},
        {20, 30, 20, 30},
        {5, 5, 5, 5},
        {15, 25, 15, 25},
        BoxSizing::ContentBox
    );

    const Size content = box.contentSize();
    const Size padding = box.paddingBoxSize();
    const Size border = box.borderBoxSize();
    const Size margin = box.marginBoxSize();

    std::cout << "Content: "
              << content.width << " x "
              << content.height << "\n";

    std::cout << "Padding box: "
              << padding.width << " x "
              << padding.height << "\n";

    std::cout << "Border box: "
              << border.width << " x "
              << border.height << "\n";

    std::cout << "Margin box: "
              << margin.width << " x "
              << margin.height << "\n";
}

void demonstrateBoxSizing() {
    std::cout << "\n=== CONTENT-BOX VS BORDER-BOX ===\n";

    const Size declared{200, 100};
    const Edges padding{20, 20, 20, 20};
    const Edges border{5, 5, 5, 5};
    const Edges margin{0, 0, 0, 0};

    BoxModel contentBox(
        declared,
        padding,
        border,
        margin,
        BoxSizing::ContentBox
    );

    BoxModel borderBox(
        declared,
        padding,
        border,
        margin,
        BoxSizing::BorderBox
    );

    const Size contentBoxResult = contentBox.borderBoxSize();
    const Size borderBoxResult = borderBox.borderBoxSize();

    std::cout << "content-box border dimensions: "
              << contentBoxResult.width << " x "
              << contentBoxResult.height << "\n";

    std::cout << "border-box border dimensions:   "
              << borderBoxResult.width << " x "
              << borderBoxResult.height << "\n";

    std::cout
        << "The same declared width can therefore represent different "
        << "content areas.\n";
}

void demonstrateOverflow() {
    std::cout << "\n=== OVERFLOW ===\n";

    const Size content{800, 600};
    const Size available{500, 400};

    const OverflowResult result =
        calculateOverflow(content, available);

    std::cout << "Horizontal overflow: "
              << result.horizontal << " px\n";

    std::cout << "Vertical overflow: "
              << result.vertical << " px\n";

    for (OverflowMode mode : {
        OverflowMode::Visible,
        OverflowMode::Hidden,
        OverflowMode::Clip,
        OverflowMode::Scroll,
        OverflowMode::Auto
    }) {
        std::cout << "Mode: "
                  << overflowDescription(mode)
                  << "\n";
    }
}

void demonstrateResponsiveSizing() {
    std::cout << "\n=== RESPONSIVE SIZING ===\n";

    const std::vector<double> viewports{
        320, 480, 768, 1024, 1440, 1920
    };

    for (double viewport : viewports) {
        const double width =
            responsiveWidth(
                viewport,
                0.80,
                280,
                900
            );

        std::cout << std::fixed
                  << std::setprecision(1)
                  << "Viewport "
                  << viewport
                  << " px -> component "
                  << width
                  << " px\n";
    }
}

void demonstrateMarginCollapse() {
    std::cout << "\n=== MARGIN COLLAPSE ===\n";

    const double first = 30;
    const double second = 20;

    std::cout << "First margin:  "
              << first << " px\n";

    std::cout << "Second margin: "
              << second << " px\n";

    std::cout << "Simplified collapsed result: "
              << collapsePositiveVerticalMargins(first, second)
              << " px\n";
}

// ============================================================================
// Verification
// ============================================================================

void expectNear(
    double actual,
    double expected,
    double tolerance,
    const std::string& description
) {
    if (std::abs(actual - expected) > tolerance) {
        std::ostringstream message;

        message
            << description
            << ": expected "
            << expected
            << ", received "
            << actual;

        throw std::runtime_error(message.str());
    }
}

void runTests() {
    std::cout << "\n=== VERIFICATION TESTS ===\n";

    BoxModel contentBox(
        {200, 100},
        {20, 20, 20, 20},
        {5, 5, 5, 5},
        {0, 0, 0, 0},
        BoxSizing::ContentBox
    );

    Size contentBoxSize = contentBox.borderBoxSize();

    expectNear(
        contentBoxSize.width,
        250,
        0.001,
        "content-box width"
    );

    expectNear(
        contentBoxSize.height,
        150,
        0.001,
        "content-box height"
    );

    BoxModel borderBox(
        {200, 100},
        {20, 20, 20, 20},
        {5, 5, 5, 5},
        {0, 0, 0, 0},
        BoxSizing::BorderBox
    );

    Size borderBoxSize = borderBox.borderBoxSize();

    expectNear(
        borderBoxSize.width,
        200,
        0.001,
        "border-box width"
    );

    expectNear(
        borderBoxSize.height,
        100,
        0.001,
        "border-box height"
    );

    OverflowResult overflow =
        calculateOverflow({600, 500}, {400, 300});

    expectNear(
        overflow.horizontal,
        200,
        0.001,
        "horizontal overflow"
    );

    expectNear(
        overflow.vertical,
        200,
        0.001,
        "vertical overflow"
    );

    expectNear(
        collapsePositiveVerticalMargins(30, 20),
        30,
        0.001,
        "positive margin collapse"
    );

    std::cout << "All tests passed.\n";
}

// ============================================================================
// Error handling demonstration
// ============================================================================

void demonstrateFailureConditions() {
    std::cout << "\n=== FAILURE CONDITIONS ===\n";

    try {
        BoxModel invalid(
            {100, 100},
            {60, 60, 60, 60},
            {10, 10, 10, 10},
            {0, 0, 0, 0},
            BoxSizing::BorderBox
        );

        // The declared 100px width cannot contain 120px padding plus borders.
        invalid.contentSize();

        std::cout << "Unexpected: invalid box accepted.\n";
    }
    catch (const std::exception& error) {
        std::cout << "Handled invalid border-box: "
                  << error.what()
                  << "\n";
    }

    try {
        BoxModel valid(
            {300, 150},
            {20, 20, 20, 20},
            {1, 1, 1, 1},
            {-10, 10, -10, 10},
            BoxSizing::BorderBox
        );

        // Negative margins are legal, so construction succeeds.
        Size result = valid.marginBoxSize();

        std::cout << "Negative-margin example: "
                  << result.width
                  << " x "
                  << result.height
                  << " px\n";
    }
    catch (const std::exception& error) {
        std::cout << "Unexpected failure: "
                  << error.what()
                  << "\n";
    }
}

// ============================================================================
// Main
// ============================================================================

} // namespace cssbox

int main() {
    try {
        std::cout
            << "CSS Box Model Technical Case Study\n"
            << "C++17 implementation\n";

        cssbox::demonstrateBasicBoxModel();
        cssbox::demonstrateBoxSizing();
        cssbox::demonstrateOverflow();
        cssbox::demonstrateResponsiveSizing();
        cssbox::demonstrateMarginCollapse();

        std::cout << "\n=== INDUSTRY-STYLE COMPONENT ===\n";

        cssbox::ProductCard card(
            "Technical Product Card",
            768
        );

        card.printReport();

        cssbox::demonstrateFailureConditions();
        cssbox::runTests();

        std::cout << "\nProgram completed successfully.\n";
    }
    catch (const std::exception& error) {
        std::cerr
            << "Fatal error: "
            << error.what()
            << "\n";

        return 1;
    }

    return 0;
}
